src/aspire/basis/fle_2d.py

```diff
@@ -0,0 +1,746 @@
+            dtype=np.complex128,
+        )
+
+        betas = np.swapaxes(betas, 0, 2)
+        betas = betas.reshape(-1, self.num_radial_nodes * num_img)
+        betas = self.r2c_nus @ betas
+        betas = betas.reshape(-1, self.num_radial_nodes, num_img)
+        betas = np.swapaxes(betas, 0, 2)
+
+        tmp[:, :, self.nus] = np.conj(betas)
+        z = fft.ifft(tmp, axis=2)
+
+        return z
+
+    def _step1(self, z):
+        """
+        Adjoint of _step1_t and final step of the forward transformation (coefficients
+            to images).
+        Performs the NUFFT on Fourier-space images to compute real-space images.
+        """
+        num_img = z.shape[0]
+        z = z[:, :, : self.num_angular_nodes // 2].reshape(num_img, -1)
+        im = anufft(
+            z,
+            np.stack((self.grid_x, self.grid_y)),
+            (self.nres, self.nres),
+            epsilon=self.epsilon,
+        )
+        im = im + np.conj(im)
+        im = np.real(im)
+        im = im.reshape(num_img, self.nres, self.nres)
+        im[:, self.radial_mask] = 0
+
+        return im
+
+    def create_dense_matrix(self):
+        """
+        Directly computes the transformation matrix from Cartesian coordinates to
+        FLE coordinates without any shortcuts.
+        :return: A NumPy array of size `(self.nres**2, self.count)` containing the matrix
+            entries.
+        """
+        # See Eqns. 3 and 4, Section 1.2
+        ts = np.arctan2(self.ys, self.xs)
+
+        B = np.zeros((self.nres, self.nres, self.count), dtype=np.complex128, order="F")
+        for i in range(self.count):
+            B[:, :, i] = self.basis_functions[i](self.rs, ts) * self.h
+        B = B.reshape(self.nres**2, self.count)
+        B = transform_complex_to_real(np.conj(B), self.ells)
+        B = B.reshape(self.nres**2, self.count)
+        if self.match_fb:
+            B[:, self.flip_sign_indices] *= -1.0
+            B = B[:, self.fb_compat_indices]
+        return B
+
+    def lowpass(self, coeffs, bandlimit):
+        """
+        Apply a low-pass filter to FLE coefficients `coeffs` with threshold `bandlimit`.
+        :param coeffs: A NumPy array of FLE coefficients, of shape (num_images, self.count)
+        :param bandlimit: Integer bandlimit (max frequency).
+        :return: Band-limited coefficient array.
+        """
+        if len(coeffs.shape) == 1:
+            coeffs = coeffs.reshape((1, coeffs.shape[0]))
+        assert (
+            len(coeffs.shape) == 2
+        ), "Input a stack of coefficients of dimension (num_images, self.count)."
+        assert (
+            coeffs.shape[1] == self.count
+        ), "Number of coefficients must match self.count."
+
+        k = self.count - 1
+        for _ in range(self.count):
+            if self.bessel_zeros[k] / (np.pi) > (bandlimit - 1) // 2:
+                k = k - 1
+        coeffs[:, k + 1 :] = 0
+
+        return coeffs
+
+    def rotate(self, coeffs, theta):
```

Comment:
`Basis.rotate` does work on these coefficients! Because it depends on the `signs_indices`, which get modified in FLE (`fle_ells_sign`), it just took splitting the signs into two versions. I would think in principle there should be a way to avoid that weird sign conversion. I'll look into it.

For now, going to make this special rotate method private and test the `Basis` one.